src/temp-alarm.ino

```diff
@@ -75,24 +102,97 @@ unsigned long elapsed() {
 
 #define LEFT_PADDING 3
 
-bool keydown = false;
-char currentKey = NULL;
+// byte currentKey = NULL;
 unsigned long keydownTime = 0;
 #define KEYDOWN_DECAY 250
 
+// interrupt callback
+void keyPressed() {
+    keyDown = true;
+}
+
+uint8_t processKey() {
+    static uint8_t lastKey;
+    uint8_t currentKey = readKey();
+
+    if (currentKey) {
+        lastKey = currentKey;
+    } else {
+        lastKey = 0;
+    }
+    return currentKey;
+}
+
+uint8_t readKey() {
+    // TODO
+    uint8_t z = 0;
+    write595(0b11111111);
+    z = read165();
+    if (z == 0) {
+        return 0;
+    }
+
+    uint8_t key = 1;
+    for (uint8_t i = 1; i <= 0b1000; i <<= 1) {
+        write595(i);
+        z = read165();
+
+        if (z != 0) {
+            break;
+        }
+        key += 4;
+    }
+
+    if (z == 0) {
+        return 0;
+    }
+
+    for (uint8_t col = 0; col < 4; col++) {
+        if ((1<<col) == z) {
+            return key + col;
+        }
+    }
+
+    return 0;
+}
+
+void write595(uint8_t b) {
+    digitalWrite(PIN_SS, LOW); // enable 595
+    asm("nop\n nop\n"); // short delay
+    SPI.transfer(b); // send byte
+    digitalWrite(PIN_SS, HIGH); // disable 595
+}
+
+uint8_t read165() {
+    digitalWrite(PIN_LOAD, LOW); // pulse load
+    asm("nop\n nop\n");

```

Comment:
Is this a clock-cycle buffer with specific timing? A comment might be useful.

Reply:
just needs to be a short delay. I saw someone do this instead of using `delay`